### ISSUE TITLE
[basic.lval] p5 An id-expression names the reference denotes the reference 

### ISSUE THREAD

**Issue by xmh0511** [cplusplus/CWG#44]
Full name of submitter (unless configured in github; will be published with the issue): Jim X

[basic.lval] p5 says
> The result of a glvalue is the entity **denoted by the expression**.

Consider this example:
````cpp
int a = 0;
int& rf = a;
rf;
````
In this case, `rf` is an unqualified-id denoting the reference, per [expr.prim.id.unqual] p3 and [basic.pre] p5. whilst `rf` is also an lvalue as per [expr.prim.id.unqual] p3. According to [basic.lval] p5, the result of lvalue `rf` is the reference. However, a `glvalue`  either denotes an object or a function, the result of the lvalue `rf` should intend to be the object denoted by `a`.  Maybe, we should use "designate" in place of "denote" in [basic.lval] p5, as the following: 
> The result of a glvalue is the entity designated by the expression.  

Everywhere, we should clearly define what entity the expression designates. Such as, we have defined what entity an unqualified-id introduced by a reference declaration **designates** in [expr.type] p1
> The expression **designates** the object or function ~denoted by the reference~ to which the reference is bound.

Such as changing [expr.unary.op] p1 as the following:
> the result is an lvalue ~referring to~ **designating** the object or function to which the expression points.

**Comment by jensmaurer** [comment]
A plethora of verbs appear to be in use to talk about the object to which a glvalue or a reference refers.

[conv.lval] p3 uses "refers" and "indicated".

[basic.lval] p5 and p11 uses "denotes"

We should use a consistent verb form here.

**Comment by jensmaurer** [comment]
It seems "denotes" should be used for "the result of name lookup", e.g. a name denotes a variable or a function. See basic.pre p5. We should pick another verb for references and glvalues.

**Comment by xmh0511** [comment]
> It seems "denotes" should be used for "the result of name lookup", e.g. a name denotes a variable or a function. See basic.pre p5.

Agree! "denote" is correctly used in [[expr.prim.id.unqual]](https://eel.is/c++draft/expr.prim.id.unqual) and [[expr.prim.id.qual]](https://eel.is/c++draft/expr.prim.id.qual). First, [basic.pre] p5 says
> An entity E is **denoted by the name** (if any) that is introduced by a declaration of E    

The name could be an unqualified name or a qualified name. 

[expr.prim.id.unqual] p3 says
> The result is the entity denoted by the [unqualified-id](https://eel.is/c++draft/expr.prim.id.unqual#nt:unqualified-id) ([[basic.lookup.unqual]](https://eel.is/c++draft/basic.lookup.unqual))[.](https://eel.is/c++draft/expr.prim.id.unqual#3.sentence-1)


[qualified-id] p5 says
> The result of a [qualified-id](https://eel.is/c++draft/expr.prim.id.qual#nt:qualified-id) Q is the entity it denotes ([[basic.lookup.qual]](https://eel.is/c++draft/basic.lookup.qual))[.](https://eel.is/c++draft/expr.prim.id.qual#5.sentence-1)

[qualified-id] p5 might be clearer after adopting https://github.com/cplusplus/draft/issues/5346

**Comment by jensmaurer** [comment]
The problem is that we use /qualified-id/ and /unqualified-id/ both for the expression / glvalue sense (which object are we talking about?) and in the /identifier/ sense (which variable or data member are we talking about?) Those two meanings need to get a clear divorce, and the right approach is to move the variable / data member interpretation into [basic], and only leave the glvalue / expression interpretation in [expr].

**Comment by xmh0511** [comment]
> The problem is that we use qualified-id and unqualified-id both for the expression / glvalue sense (which object are we talking about?) and in the identifier sense (which variable or data member are we talking about?) 

IIUC, do you mean that the name of a variable or data member can be associated with several objects? I want to divorce objects from variables. "variable" should have a pure meaning that its name could be used as an lvalue. For the problem that which object is designated/referred to by the **name**, I think it is determined by the evaluation of the `glvalue`. 

***
In the current, the entity is an object, function, or something else specified in [basic.pre]. 

> The result is the entity denoted by the unqualified-id ([basic.lookup.unqual]).

I construe this rule as that the name is used to denote (conceptually)objects rather than the concrete object. Because [basic.lookup] only associates the name with the found declaration, it cannot determine which object the name refers to.

**Comment by jensmaurer** [comment]
> do you mean that the name of a variable or data member can be associated with several objects?

Yes. Think about a local variable in a recursive function: you get a different local variable for each invocation.
